daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Runner.scala

```diff
@@ -283,6 +283,33 @@ object Runner {
 class Runner(compiledPackages: CompiledPackages, script: Script.Action, timeMode: ScriptTimeMode)
     extends StrictLogging {
 
+  // We overwrite the definition of fromLedgerValue with an identity function.
+  // This is a type error but Speedy doesn’t care about the types and the only thing we do
+  // with the result is convert it to ledger values/record so this is safe.
+  private val extendedCompiledPackages = {
+    val fromLedgerValue: PartialFunction[SDefinitionRef, SExpr] = {
+      case LfDefRef(id) if id == script.scriptIds.damlScript("fromLedgerValue") =>
+        SEMakeClo(Array(), 1, SELocA(0))
+    }
+    new CompiledPackages(Compiler.FullStackTrace, Compiler.NoProfile) {
+      override def getPackage(pkgId: PackageId): Option[Package] =
+        compiledPackages.getPackage(pkgId)
+      override def getDefinition(dref: SDefinitionRef): Option[SExpr] =
+        fromLedgerValue.andThen(Some(_)).applyOrElse(dref, compiledPackages.getDefinition)
+      // FIXME: avoid override of non abstract method
+      override def packages: PartialFunction[PackageId, Package] = compiledPackages.packages
+      override def packageIds: Set[PackageId] = compiledPackages.packageIds
+      // FIXME: avoid override of non abstract method
+      override def definitions: PartialFunction[SDefinitionRef, SExpr] =
+        fromLedgerValue.orElse(compiledPackages.definitions)
+      override def packageLanguageVersion: PartialFunction[PackageId, LanguageVersion] =
+        compiledPackages.packageLanguageVersion
+    }
+  }
+
+  val machine =
+    Speedy.Machine.fromPureSExpr(extendedCompiledPackages, script.expr, onLedger = false)
```

Comment:
This means the machine is now persisted across multiple runs, right? Does this mean we need to reset it after runs, e.g. between test cases in `daml test-script`?

Reply:
An indvidual `Runner` is specific to a single script. I’m not aware of any case where we call `run` multiple for a single script or reuse `Runner` at all so this doesn’t really matter. But I agree that the API here is confusing and doesn’t enforce that this is used correctly and we should fix this. I’ve changed it to return the machine from `runWithClients` now. It feels a bit weird to return `(Machine, Future[_])` but it does the trick and avoids this. We should probably also kill the `Runner`constructer since we never construct a runner without running it immediately.